arch/riscv/core/isr.S

```diff
@@ -163,6 +163,14 @@ SECTION_FUNC(exception.entry, _isr_wrapper)
 	lr t0, ___cpu_t_current_OFFSET(s0)
 	lr tp, _thread_offset_to_tls(t0)
 
+	/* Make sure global pointer is sane */
+#ifdef CONFIG_RISCV_GP
+	.option push
+	.option norelax
+	la gp, __global_pointer$
```

Comment:
Just some notes

It looks like GP is specified as neither callee saved nor caller saved in the riscv spec.

OpenSBI backs it up and restores GP in their handler.

https://www.openeuler.org/en/blog/zhangxinhao/risc-v/2020-11-28-RISC-V.html

I think we're under no obligation to restore it or back it up.

Some bare metal environments might use it.

https://five-embeddev.com/quickref/global_pointer.html


Is the point here to illustrate that we _can_ use it from within uninterruptible context?

Reply:
GCC seems to never save GP: https://github.com/gcc-mirror/gcc/blob/master/gcc/config/riscv/riscv.cc#L7368-L7370

Reply:
It's probably smart to support at least one convention for consistency in the kernel.

I can also see kernel-side code potentially using  GP as well, but if the preempted threads value is destroyed by an ISR with higher priority, I would certainly want to have it back the way I expected.

Maybe this below? Is this what we already do?

* back up the old GP, so the preempted context can still have a consistent register view or return
* do what we do here, and set it to a default value on entering context (ISR can adjust however)
* restore on return from interrupt

@npitre, let me know if this aligns with your suggestion as well. I understand it's adding one more register to push / pop, but it's maybe worth it?